ark/utils/google_drive_utils.py

```diff
@@ -0,0 +1,364 @@
+        return None
+
+    def write(self, data, overwrite=False):
+        """ Writes file data or data from file path to provided Drive file.
+
+        Args:
+            data (str or BytesIO):
+                If BytesIO, data is directly uploaded.  If str, data is assumed to be a filepath
+                and data is attempted to be read from the file.
+            overwrite (bool):
+                If a file already exists at the path_string and overwrite is False, no data is
+                written.  Otherwise, the existing file on Drive is updated/overwritten.
+        """
+        if self._service_check() or self.parent_id_map is None:
+            return
+
+        # TODO: implement this
+        return
+
+    def lsfiles(self):
+        """ List all non-directory files at path_string in Drive
+        """
+        if self._service_check() or self.fileID is None:
+            return None
+
+        filenames = []
+
+        global SERVICE
+        page_token = None
+        while True:
+            response = SERVICE.files().list(
+                q=f"(('{self.fileID}' in parents) and ({_FILE_MIME_CHECK}))",
+                spaces='drive',
+                fields='nextPageToken, files(name, mimeType, shortcutDetails(targetMimeType))',
+                pageToken=page_token).execute()
+
+            for file in response.get('files', []):
+                if file.get('shortcutDetails', None) is not None:
+                    mimeType = file.get('shortcutDetails').get('targetMimeType')
+                    if mimeType != _FOLDER_MIME:
+                        filenames.append(file.get('name'))
+                else:
+                    filenames.append(file.get('name'))
+
+            page_token = response.get('nextPageToken', None)
+            if page_token is None:
+                break
+
+        return filenames
+
+    def lsdirs(self):
+        """ List all directories at path_string in Drive
+        """
+        if self._service_check() or self.fileID is None:
+            return None
+
+        dirnames = []
+
+        global SERVICE
+        page_token = None
+        while True:
+            response = SERVICE.files().list(
+                q=f"(('{self.fileID}' in parents) " +
+                  f"and (({_FOLDER_MIME_CHECK}) or ({_SHORTCUT_MIME_CHECK})))",
+                spaces='drive',
+                fields='nextPageToken, files(name, shortcutDetails(targetMimeType))',
+                pageToken=page_token).execute()
+
+            for file in response.get('files', []):
+                if file.get('shortcutDetails', None) is not None:
+                    mimeType = file.get('shortcutDetails').get('targetMimeType')
+                    if mimeType == _FOLDER_MIME:
+                        dirnames.append(file.get('name'))
+
+                dirnames.append(file.get('name'))
+
+            page_token = response.get('nextPageToken', None)
+            if page_token is None:
+                break
```

Comment:
Can we put this code in a separate function? Seems like similar logic is repeated in `lsfiles`.